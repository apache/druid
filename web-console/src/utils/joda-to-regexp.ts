/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Refer to https://www.joda.org/joda-time/key_format.html
const TEXT = '\\w+';
const NUMBER_2_DIGIT = '[0-9]{2}';
const NUMBER_4_DIGIT = '[0-9]{4}';
const JODA_FRAGMENT_TO_REG_EXP_STRING: Record<string, string> = {
  C: '[0-9]{1,2}',
  CC: NUMBER_2_DIGIT,
  YY: NUMBER_2_DIGIT,
  YYYY: NUMBER_4_DIGIT,

  xx: NUMBER_2_DIGIT,
  xxxx: NUMBER_4_DIGIT,
  w: '[0-9]{1,2}',
  ww: NUMBER_2_DIGIT,
  e: '[0-7]',
  E: TEXT,
  EEEE: TEXT,

  yy: NUMBER_2_DIGIT,
  yyyy: NUMBER_4_DIGIT,
  D: '[0-9]{1,3}',
  DD: '[0-9]{2,3}',
  DDD: '[0-9]{3}',
  M: '(?:1[0-2]|[1-9])',
  MM: '(?:1[0-2]|0[1-9])',
  MMM: TEXT,
  MMMM: TEXT,
  d: '(?:3[0-1]|[12][0-9]|[1-9])',
  dd: '(?:3[0-1]|[12][0-9]|0[1-9])',

  a: '[ap]m',
  K: '(?:1[01]|[0-9])',
  KK: '(?:1[01]|0[0-9])',
  h: '(?:1[0-2]|[1-9])',
  hh: '(?:1[0-2]|0[1-9])',

  H: '(?:2[0-3]|1[0-9]|[0-9])',
  HH: '(?:2[0-3]|1[0-9]|0[0-9])',
  k: '(?:2[0-4]|1[0-9]|[1-9])',
  kk: '(?:2[0-4]|1[0-9]|0[1-9])',
  m: '(?:[1-5][0-9]|[0-9])',
  mm: '[0-5][0-9]',
  s: '(?:[1-5][0-9]|[0-9])',
  ss: '[0-5][0-9]',
  S: '[0-9]{1,3}',
  SS: '[0-9]{2,3}',
  SSS: '[0-9]{3}',
  z: TEXT,
  Z: TEXT
};

export function jodaFormatToRegExp(jodaFormat: string): RegExp {
  const regExpStr = jodaFormat.replace(/([a-zA-Z])\1{0,3}/g, jodaPart => {
    const re = JODA_FRAGMENT_TO_REG_EXP_STRING[jodaPart];
    if (!re) throw new Error(`could not convert ${jodaPart} to RegExp`);
    return re;
  });
  return new RegExp(`^${regExpStr}$`, 'i');
}
