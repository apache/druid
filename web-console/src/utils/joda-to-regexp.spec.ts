/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { jodaFormatToRegExp } from './joda-to-regexp';

describe('jodaFormatToRegExp', () => {
  it('works for common formats', () => {
    expect(jodaFormatToRegExp('d/M/yyyy').toString()).toMatchInlineSnapshot(
      `"/^(?:3[0-1]|[12][0-9]|[1-9])\\\\/(?:1[0-2]|[1-9])\\\\/[0-9]{4}$/i"`,
    );

    expect(jodaFormatToRegExp('MM/dd/YYYY').toString()).toMatchInlineSnapshot(
      `"/^(?:1[0-2]|0[1-9])\\\\/(?:3[0-1]|[12][0-9]|0[1-9])\\\\/[0-9]{4}$/i"`,
    );

    expect(jodaFormatToRegExp('M/d/YY').toString()).toMatchInlineSnapshot(
      `"/^(?:1[0-2]|[1-9])\\\\/(?:3[0-1]|[12][0-9]|[1-9])\\\\/[0-9]{2}$/i"`,
    );

    expect(jodaFormatToRegExp('d-M-yyyy hh:mm:ss a').toString()).toMatchInlineSnapshot(
      `"/^(?:3[0-1]|[12][0-9]|[1-9])-(?:1[0-2]|[1-9])-[0-9]{4} (?:1[0-2]|0[1-9]):[0-5][0-9]:[0-5][0-9] [ap]m$/i"`,
    );

    expect(jodaFormatToRegExp('MM/dd/YYYY hh:mm:ss a').toString()).toMatchInlineSnapshot(
      `"/^(?:1[0-2]|0[1-9])\\\\/(?:3[0-1]|[12][0-9]|0[1-9])\\\\/[0-9]{4} (?:1[0-2]|0[1-9]):[0-5][0-9]:[0-5][0-9] [ap]m$/i"`,
    );

    expect(jodaFormatToRegExp('YYYY-MM-dd HH:mm:ss').toString()).toMatchInlineSnapshot(
      `"/^[0-9]{4}-(?:1[0-2]|0[1-9])-(?:3[0-1]|[12][0-9]|0[1-9]) (?:2[0-3]|1[0-9]|0[0-9]):[0-5][0-9]:[0-5][0-9]$/i"`,
    );

    expect(jodaFormatToRegExp('YYYY-MM-dd HH:mm:ss.S').toString()).toMatchInlineSnapshot(
      `"/^[0-9]{4}-(?:1[0-2]|0[1-9])-(?:3[0-1]|[12][0-9]|0[1-9]) (?:2[0-3]|1[0-9]|0[0-9]):[0-5][0-9]:[0-5][0-9].[0-9]{1,3}$/i"`,
    );
  });

  it('matches dates when needed', () => {
    expect(jodaFormatToRegExp('d-M-yyyy hh:mm:ss a').test('26-4-1986 01:23:40 am')).toEqual(true);
    expect(jodaFormatToRegExp('YYYY-MM-dd HH:mm:ss.S').test('26-4-1986 01:23:40 am')).toEqual(
      false,
    );
  });
});
