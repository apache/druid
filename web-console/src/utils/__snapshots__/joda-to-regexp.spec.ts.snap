// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jodaFormatToRegExp works for common formats 1`] = `"/^(?:3[0-1]|[12][0-9]|[1-9])\\\\/(?:1[0-2]|[1-9])\\\\/[0-9]{4}$/i"`;

exports[`jodaFormatToRegExp works for common formats 2`] = `"/^(?:1[0-2]|0[1-9])\\\\/(?:3[0-1]|[12][0-9]|0[1-9])\\\\/[0-9]{4}$/i"`;

exports[`jodaFormatToRegExp works for common formats 3`] = `"/^(?:1[0-2]|[1-9])\\\\/(?:3[0-1]|[12][0-9]|[1-9])\\\\/[0-9]{2}$/i"`;

exports[`jodaFormatToRegExp works for common formats 4`] = `"/^(?:3[0-1]|[12][0-9]|[1-9])-(?:1[0-2]|[1-9])-[0-9]{4} (?:1[0-2]|0[1-9]):[0-5][0-9]:[0-5][0-9] [ap]m$/i"`;

exports[`jodaFormatToRegExp works for common formats 5`] = `"/^(?:1[0-2]|0[1-9])\\\\/(?:3[0-1]|[12][0-9]|0[1-9])\\\\/[0-9]{4} (?:1[0-2]|0[1-9]):[0-5][0-9]:[0-5][0-9] [ap]m$/i"`;

exports[`jodaFormatToRegExp works for common formats 6`] = `"/^[0-9]{4}-(?:1[0-2]|0[1-9])-(?:3[0-1]|[12][0-9]|0[1-9]) (?:2[0-3]|1[0-9]|0[0-9]):[0-5][0-9]:[0-5][0-9]$/i"`;

exports[`jodaFormatToRegExp works for common formats 7`] = `"/^[0-9]{4}-(?:1[0-2]|0[1-9])-(?:3[0-1]|[12][0-9]|0[1-9]) (?:2[0-3]|1[0-9]|0[0-9]):[0-5][0-9]:[0-5][0-9].[0-9]{1,3}$/i"`;
